--- videzzo_fork.c	2024-02-06 03:49:45.104116784 +0000
+++ videzzo_fork.c	2024-02-06 04:40:50.187043020 +0000
@@ -6,6 +6,10 @@
  * This work is licensed under the terms of the GNU GPL, version 2 or later.
  */
 
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
 #include "videzzo_fork.h"
 
 //
@@ -32,4 +36,4 @@
     memcpy(&__FUZZ_COUNTERS_START, copy,
            &__FUZZ_COUNTERS_END - &__FUZZ_COUNTERS_START);
     free(copy);
-}
\ No newline at end of file
+}
