--- include/qemu/cutils.h.origin	2024-10-02 04:08:55.812917568 +0000
+++ include/qemu/cutils.h	2024-10-02 04:09:09.905203022 +0000
@@ -241,10 +241,13 @@
 int qemu_pstrcmp0(const char **str1, const char **str2);
 
 /* Find program directory, and save it for later usage with
- * get_relocated_path().
+ * qemu_get_exec_dir().
  * Try OS specific API first, if not working, parse from argv0. */
 void qemu_init_exec_dir(const char *argv0);
 
+/* Get the saved exec dir.  */
+const char *qemu_get_exec_dir(void);
+
 /**
  * get_relocated_path:
  * @dir: the directory (typically a `CONFIG_*DIR` variable) to be relocated.
